_overviews/tutorials/binary-compatibility-for-library-authors.md

```diff
@@ -0,0 +1,215 @@
+It works by comparing the class files of two provided JARs and report any binary incompatibilities found.
+
+By incorporating [MiMa SBT plugin](https://github.com/typesafehub/migration-manager/wiki/Sbt-plugin) into your SBT build, you can easily check whether 
+you have accidentally introduced binary incompatible changes. Detailed instruction on how to use the SBT plugin can be found in the link.
+
+We strongly encourage every library author to incorporate MiMa into their library release workflow.
+
+## Evolving code without breaking binary compatibility
+
+Binary compatibility breakages can often be avoided through careful use of certain Scala features as well as some techniques you can apply when modifying code.
+
+Some language features may break binary compatibility:
+
+* Default parameter values for methods or classes
+* Case classes
+* Default methods on traits (doesn't cause breakages since 2.12)
+
+Techniques you can use to avoid breaking binary compatibility:
+
+* Annotate public method's return type explicitly
+* Mark methods as package private when you want to remove a method or modify its signature
+* Don't turn on inlining for methods from dependencies
+
+For brevity of this guide, detailed explanation and runnable code examples can be found in [Binary Compatibility Code Examples & Explanation](https://github.com/jatcwang/binary-compatibility-guide).
+
+Again, we recommend using MiMa to double check that you have not broken binary compatibility after making changes.
+
+## Versioning Scheme - Communicating compatibility breakages
+
+Library authors use versioning schemes to communicate compatibility guarantees between library releases to their users. Versioning schemes like [Semantic Versioning](http://semver.org/)(SemVer) allow
+users to easily reason about the impact of updating a library, without needing to read the detailed release note.
+
+In the following section, we will outline a versioning scheme based on Semantic Versioning that we **strongly encourage** you to adopt for your libraries. The rules listed below are **in addition** to 
+Semantic Versioning v2.0.0.
+
+### Recommended Versioning Scheme
+
+* If backward **binary compatibility** is broken, **major version number** must be increased
+* If backward **source compatibility** is broken, **minor version number** must be increased
+* A change in **patch version number** signals **no binary nor source incompatibility**. According to SemVer, patch versions should contain only bug fixes that fix incorrect behavior so major behavioral
+change in method/classes should result in a minor version bump.
+* When major version is `0`, a minor version bump **may contain both source and binary breakages**
+* Some libraries may take a harder stance on maintaining source compatibility, bumping the major version number for ANY source incompatibility even if they are binary compatible
```

Comment:
This is not a recommendation, just a clarification, so I think it should be in an independent paragraph. The gist would be: "being stricter than this scheme is possible and encouraged when possible, though most of the times it's not required; here's how...".

Reply:
Bumping the major version for ANY source incompatibility is probably nonsense. If you really mean *ANY* (and I guess you mean it, because it's uppercase), then adding a public method means you need to bump MAJOR. I don't any library will ever be that strict.

Actually, I would argue that a library doing this is doing a *disservice* to its community, as it has no means to effectively convey *binary incompatible changes* anymore in a way that is independent from source incompatible changes. This means that every single new public method breaks the ecosystem, since I cannot know from the version number whether I am moving into dependency hell or not.

Reply:
MAJOR could be bumped for "severe" source breaking changes (for some definition of "severe"). Removing public APIs could be considered "severe", for example. But we cannot give any objective metric about this.

Reply:
I think I'll remove this paragraph to avoid confusion.